# Python Basics

**CS1302 Introduction to Computer Programming**
___

In [2]:
import sys
from ipywidgets import interact

%reload_ext divewidgets

**Content**
- Syntax, Comments, Literal, Variable, Function, Identifier
- Basic data types
- string formatting

## Syntax

- In computer science, the `syntax` is a set of rules for programming. In other words, it means using legal structures and commands that a computer can interpret.
- Human language is used to communicate with people. Programming language is used to communicate with machine. If we don't follow the correct syntax, the computer can't run the code (syntax error). See example below.

In [2]:
# print() is a function to print a message out on the screen
print("Hello, World!" #this line has error because it misses one ) in the end, so the computer raises SyntaxError

SyntaxError: incomplete input (1712927172.py, line 2)

In [3]:
print("Hello, World!") #this code is correct, and it will print "Hello, World" on the screen
print("Something I'd like to print")

Hello, World!
Something I'd like to print


## Comments

- In the above example, the texts after hash key # is called comments

- Why we need comments?
    - Comments can be used to explain Python code.
    - Comments can be used to make the code more readable.
    - Comments can be used to prevent execution when testing code.

### Create a single line comment
Comments starts with a #, and Python will ignore them:

In [4]:
#This program prints "Hello, World!"
print("Hello, World!")

Hello, World!


Comments can be placed at the end of a line, and Python will ignore the rest of the line:

In [5]:
print("Hello, World!")  #This program prints "Hello, World!"

Hello, World!


A comment does not have to be text that explains the code, it can also be used to prevent Python from executing code:

In [6]:
print("Hello, World!")
#print("Cheers, Mate!")

Hello, World!


### Create multi-line comments

Python does not really have a syntax for multi-line comments. To add a multi-line comment you could insert a # for each line:

In [ ]:
#This program
#prints
#"Hello, World!"
print("Hello, World!")

Or, not quite as intended, you can use a multiline string.

Since Python will ignore string literals that are not assigned to a variable, you can add a multiline string (triple quotes) in your code, and place your comment inside it:

In [19]:
"""
This is a comment
written in
more than just one line
This is more comments
This is another comment
"""
print("Hello, World!")

Hello, World!


## Literal

literal is a notation for representing a fixed value in source code. They can also be defined as raw value or data given in variables or constants. 

Python has different types of literals, such as:
1. Integer literal
    - `78` is an integer literal
2. Floating point literal
    - `23.32` is a floating point literal 
3. character literal
    - `'a'` is a character literal
4. String literal
    - `"hello"` is a string literal

In [21]:
x = 78 #78 is integer literal
y = 23.32 #23.32 is floating point literal
z = "a" #"a" is character literal
text = "hello" #string literal
x = 100

## Variable

- A variable is just a name to store values. This means that when you create a variable you reserve some space in memory.

- Based on the data type of a variable, the computer allocates memory and decides what can be stored in the reserved memory. Therefore, by assigning different data types to variables, you can store integers, decimals or characters in these variables.

### how to create variables
- Python variables do not need explicit declaration to reserve memory space. The declaration happens automatically when you assign a value to a variable. The equal sign (=) is used to assign values to variables.

- The operand to the left of the = operator is the name of the variable and the operand to the right of the = operator is the value stored in the variable. For example:

In [24]:
# how to create a variable
counter = 100        # this creates a variable counter which points to the memory where stores an integer 100
miles   = 100.23     # this creates a variable miles which points to the memory where stores a floating point number 100.23
name    = "John"     # this creates a variable name which points to the memory where stores a string "John"

#== #in Python, this operator means the equal operator

print(counter)
print(miles)
print(name)

100
100.23
John


### Object References
What is actually happening when you make a variable assignment? 

Python is a object-oriented language. In fact, every item of data in a Python program is an object of a specific type or class (object will be introduced later in this course). 

Considering the following code:

In [25]:
print(300)

300


When presented with the statement `print(300)`, the interpreter does the following:

- Creates an integer object
- Gives it the value 300
- Displays it on the screen

A Python variable is a symbolic name that is a reference or pointer to an object. Once an object is assigned to a variable, you can refer to the object by that name. But the data itself is still contained within the object.

For example:

In [27]:
n = 300

This assignment creates an integer object with the value 300 and assigns the variable n to point to that object.



<center><figure>
<a title="Unsplash" href="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable1.png"><img width="400" alt="Variable assignment" src="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable1.png"></a>
  <figcaption>Variable Assignment.</figcaption>
</figure>
</center>

Now consider the following statement:

In [28]:
m = n

What happens when it is executed? Python does not create another object. It simply creates a new symbolic name or reference, `m`, which points to the same object that `n` points to.

<center><figure>
<a title="Unsplash" href="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable2.png"><img width="600" alt="Variable assignment" src="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable2.png"></a>
  <figcaption>Multiple References to a Single Object.</figcaption>
</figure>
</center>

Next, suppose you do this:

In [ ]:
m = 400

Now Python creates a new integer object with the value 400, and `m` becomes a reference to it.

<center><figure>
<a title="Unsplash" href="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable3.png"><img width="600" alt="Variable assignment" src="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable3.png"></a>
  <figcaption>References to Separate Objects.</figcaption>
</figure>
</center>

Lastly, suppose this statement is executed next:

In [29]:
n = "foo"

Now Python creates a string object with the value "foo" and makes `n` reference that.


<center><figure>
<a title="Unsplash" href="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable4.png"><img width="600" alt="Variable assignment" src="https://www.cs.cityu.edu.hk/~weitaoxu/cs1302/variable4.png"></a>
  <figcaption>Orphaned Object.</figcaption>
</figure>
</center>

There is no longer any reference to the integer object 300. It is orphaned, and there is no way to access it.

**Is assignment the same as equality?**

No because:
- `=` is assignment operator, `x = 15` means we create a variable `x` and assign 15 to `x`
- `==` is comparison operator, `x == 15` means we compare the value of `x` with 15. It returns `True` if x is equal to 15; otherwise, it returns `False`

**Exercise** Try out the code by yourself.

In [3]:
x = 15
print(15 == x)
print(x == 15)

x = 20
print(15 == x)
print(x == 15)


#15 == x is equivalent to x == 15; but 15 = x is different from x = 15
#the code below is wrong
#15=x

x == 15
15 == x

x=15
15 = x

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (4261466296.py, line 18)

In [6]:
# is a keyword to check whether two objects are the same object or not

x = 300
y = x #this is called "aliasing" in programming
z = 300
print(x is y)
print(x is z)

True
False


Let's see the effect of assignment step-by-step:
1. Run the following cell.
1. Click `Next >` to see the next step of the execution.

In [7]:
%reload_ext divewidgets

In [8]:
%%optlite --height 400
x = 300
x = x + 1

OPTWidget(value=None, height=400, script='x = 300\nx = x + 1\n')

The following *tuple assignment* syntax can assign multiple variables in one line.

In [14]:
%%optlite --height 200
x, y, z = '15', '30', 15

OPTWidget(value=None, height=200, script="x, y, z = '15', '30', 15\n")

One can also use *chained assignment* to set different variables to the same value.

In [15]:
%%optlite --height 200
x = y = z = 0

OPTWidget(value=None, height=200, script='x = y = z = 0\n')

Variables can be deleted using `del`. Accessing a variable before assignment raises a Name error.

In [39]:
x = 5
y = 10
del x, y
x, y

NameError: name 'x' is not defined

## Function
- A `function` is a block of code which only runs when it is called.
- You can pass data, known as `parameters` or `arguments`, into a function.
- A function can return data as a result.
- `print()` is a function to print the message on the screen

**How to define a function**

In Python a function is defined using the `def` keyword:

**How to call a function**

To call a function, use the function name followed by parenthesis:

In [42]:
def m1():  #this is how to define a function
    print("Hello from a function")
    
m1()     #this is how to call a function

Hello from a function


**Arguments**

- Information can be passed into functions as arguments.

- Arguments are specified after the function name, inside the parentheses. You can add as many arguments as you want, just separate them with a comma.

- The following example has a function with one argument (fname). When the function is called, we pass along a first name, which is used inside the function to print the full name:

In [50]:
def my_function(course_name):
    print("Welcome to "+ course_name)

In [51]:
my_function("CityU")

Welcome to CityU


To know how to use a built-in function, we can add a question mark before or after the function name:

In [47]:
?print

Docstring:
print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

Prints the values to a stream, or to sys.stdout by default.
Optional keyword arguments:
file:  a file-like object (stream); defaults to the current sys.stdout.
sep:   string inserted between values, default a space.
end:   string appended after the last value, default a newline.
flush: whether to forcibly flush the stream.
Type:      builtin_function_or_method


In [14]:
print?

Docstring:
print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

Prints the values to a stream, or to sys.stdout by default.
Optional keyword arguments:
file:  a file-like object (stream); defaults to the current sys.stdout.
sep:   string inserted between values, default a space.
end:   string appended after the last value, default a newline.
flush: whether to forcibly flush the stream.
Type:      builtin_function_or_method


**import**

Some functions are defined in special libraries and will not be loaded automatically. We need to import these modules before using the functions.

For example, we can import the math module by typing the following line:

In [48]:
import math

then, we can use the values and functions defined in the math module

In [49]:
print(math.pi) #pi is a constant defined in math module
print(math.sin(1.25)) #sin() is a function defined in math module

3.141592653589793
0.9489846193555862


## Identifiers

`Identifer` is a name to represent various program elements such as variables, arrays, functions etc.


*Identifiers* such as variable names are case sensitive and follow certain rules.
- `a = 10` is different from `A = 10`

In [9]:
%%optlite --height 400
a = 10
A = 10

OPTWidget(value=None, height=400, script='a = 10\nA = 10\n')

**What is the syntax for variable names?**

1. Must start with a letter or `_` (an underscore) followed by letters, digits, or `_`.
1. Must not be a [keyword](https://docs.python.org/3.7/reference/lexical_analysis.html#keywords) reserved by python:

```python
False      await      else       import     pass
None       break      except     in         raise
True       class      finally    is         return
and        continue   for        lambda     try
as         def        from       nonlocal   while
assert     del        global     not        with
async      elif       if         or         yield
```

**Exercise** Evaluate the following cell and check if any of the rules above is violated.

```{caution}
For the code to run, you must run the initialization cell that imports ipywidgets:
````python
from ipywidgets import interact
````
```

In [18]:
@interact
def identifier_syntax(
    assignment=[
        "a-number = 15",
        "a_number = 15",
        "15 = 15",
        "_15 = 15",
        "del = 15",
        "Del = 15",
        "type = print",
        "print = type",
        "input = print",
    ]
):
    exec(assignment)
    print("Ok.")

interactive(children=(Dropdown(description='assignment', options=('a-number = 15', 'a_number = 15', '15 = 15',…

What can we learn from the above examples?

```{hint}
- `del` is a keyword and `Del` is not because identifiers are case sensitive.
- Function names such as `print`, `input`, `type`, etc., are not keywords and can be reassigned.  
  This can useful if you want to modify the default implementations without changing their source code.
```

In [60]:
print('Hello, World!')
#although print is the name of a function, it's not a keyword
x=print
x('Hello, World!') #equivalent to print('Hello, World!')

Hello, World!
Hello, World!


```{seealso}
To help make their code more readable, programmers follow additional style guides such as [PEP 8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names):
- Function names should be lowercase, with words separated by underscores as necessary to improve readability.  
- Variable names follow the same convention as function names.
- good variable names: my_age, my_weight
- bad variable names: xyz, abc123
```

## Basic data types

In programming, data type is an important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:

- Text Type:	    `str`
- Numeric Types:	`int`, `float`, `complex`
- Sequence Types:	`list`, `tuple`, `range`
- Mapping Type:	`dict`
- Set Types:	`set`, `frozenset`
- Boolean Type:	`bool`
- Binary Types:	`byte`, `bytearray`, `memoryview`

In this lecture, we only introduce the most basic data types:
- integer: represented by keyword `int`
- floating point number: represented by keyword `float`
- string: represented by keyword `str`

### Integer

**How to enter an [integer](https://docs.python.org/3/reference/lexical_analysis.html#integer-literals) in a program?**

In [61]:
15  # an integer in decimal

15

In [62]:
0b1111  # a binary number

15

In [63]:
0xF  # hexadecimal (base 16) with possible digits 0, 1,2,3,4,5,6,7,8,9,A,B,C,D,E,F

15

**Why all outputs are the same?**

- What you have entered are *integer literals*, which are integers written out literally. 
- All the literals have the same integer value in decimal.
- By default, if the last line of a code cell has a value, the jupyter notebook will store and display the value as an output. 

In [19]:
3  # not the output of this cell
4+5+6 #this is the output of this cell

15

- The last line above also has the same value, `15`.
- It is an *expression* (but not a literal) that *evaluates* to the integer value.

**Exercise** 

Enter an expression that evaluates to an integer value, as big as possible.  
(You may need to interrupt the kernel if the expression takes too long to evaluate.)

In [72]:
# YOUR CODE HERE
2**1000

```{seealso}
Is there a maximum value for an integer for Python3?
See the [documentation](https://docs.python.org/3.1/whatsnew/3.0.html#integers) about `sys.maxint`.
```

## Floating Point Numbers

Not all numbers are integers. In Enginnering, we often need to use fractions.

**How to enter fractions in a program?**

In [21]:
x = -0.1  # decimal number
y = -1.0e-1  # scientific notation
z = -1 / 10  # fraction
x, y, z, type(x), type(y), type(z)

(-0.1, -0.1, -0.1, float, float, float)

In the above code, `type()` is a python function to return the data type of the inputs.

Run the code below to see how to use `type()`

In [78]:
type?

Init signature: type(self, /, *args, **kwargs)
Docstring:     
type(object) -> the object's type
type(name, bases, dict, **kwds) -> a new type
Type:           type
Subclasses:     ABCMeta, EnumMeta, NamedTupleMeta, _TypedDictMeta, _ABC, MetaHasDescriptors, PyCStructType, UnionType, PyCPointerType, PyCArrayType, ...


**What is the type `float`?**

- `float` corresponds to the [*floating point* representation](https://en.wikipedia.org/wiki/Floating-point_arithmetic#Floating-point_numbers).  
- A `float` in stored exactly the way we write it in scientific notation: 

$$
\overbrace{-}^{\text{sign}} \underbrace{1.0}_{\text{mantissa}\kern-1em}e\overbrace{-1}^{\text{exponent}\kern-1em}=-1\times 10^{-1}
$$
- An efficient implementation is more complicated. Try the [IEEE-754 Floating Point Converter](https://www.h-schmidt.net/FloatConverter/IEEE754.html).

Special properties of `float`

`Property 1`. Python stores a [floating point](https://docs.python.org/3/library/sys.html#sys.float_info) with finite precision, which affects the check for equality. See the code below.

In [80]:
x = 10
y = (x ** (1 / 3)) ** 3
x == y

print(x)
print(y)

10
10.000000000000002


Why False? Shouldn't $(x^{\frac13})^3=x$?
- Floating point numbers have finite precisions and so  
- we should instead check whether the numbers are close enough.

How to compare floating point numbers?
- `isclose()` function
- syntax: `isclose(x,y)`
- `isclose(x,y)` function checks whether two values (i.e., x and y) are close to each other, or not. Returns `True` if the values are close, otherwise `False`.
- `isclose(x,y)` function is defined in `math` module;therefore, we need to first import `math` module, then use the function as `math.isclose(x,y)`

In [22]:
import math #isclose() function is defined in math module, so we need to import math module first
x = 10
y = (x ** (1 / 3)) ** 3

print(x == y)  #this returns False
print(math.isclose(x,y)) #this returns True  . is called member operator

False
True


`Property 2`. There is no limit or maximum value of an integer in Python 3. You can create a value as large as the machine can store. However, there is a maximum value for floating point number. Run the following code to see the maximum number in your system (no need to remember how it works).

In [23]:
# There is no maximum for an integer for Python3. 
# See https://docs.python.org/3.1/whatsnew/3.0.html#integers

In [1]:
import sys #this line imports sys module

print(sys.float_info) #this line will print the information of float data type 

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)


It cannot represent a number larger than the `max`:

In [2]:
print(sys.float_info.max * 2)

inf


`Property 3:` it may keep more decimal places than desired.

In [84]:
1/3

0.3333333333333333

**How to [round](https://docs.python.org/3/library/functions.html#round) a floating point number to the desired number of decimal places?**
- `round()` function
- syntax: `round(number, digits)`
- `number`-Required. The number to be rounded
- `digits`-Optional. The number of decimals to use when rounding the number. Default is 0

In [85]:
x = 28793.54836
print('round(x)=',round(x)) #round to ones place
print('round(x, 2)=',round(x, 2)) # round to 2 decimal places
print('round(x, 1)=',round(x, 1)) #round to 1 decimal places
print('round(x, 0)=',round(x, 0)) #round to ones place, but return a float
print('round(x, -1)=',round(x, -1)) #round to tens place
print('round(x, -2)=',round(x, -2)) #round to hundreds place

round(x)= 28794
round(x, 2)= 28793.55
round(x, 1)= 28793.5
round(x, 0)= 28794.0
round(x, -1)= 28790.0
round(x, -2)= 28800.0


In [86]:
round(2.665, 2), round(2.675, 2)

(2.67, 2.67)

**Why 2.675 rounds to 2.67 instead of 2.68?**

- The reason is complex,simply speaking it is because [the number may not be represented exactly in binary](https://docs.python.org/3/tutorial/floatingpoint.html#tut-fp-issues) (optional).

The `round` function can also be applied to an integer.

In [21]:
round(150, -2), round(250, -2)

(200, 200)

**Why 250 rounds to 200 instead of 300?**

- Python 3 implements the default rounding method in [IEEE 754](https://en.wikipedia.org/w/index.php?title=IEEE_754#Rounding_rules). (optional)

## Strings

**How to enter a string in a program?**

- Strings in python are surrounded by either single quotation marks, double quotation marks, three single quotation marks, or three double quotation marks.
- 'hello', "hello", '''hello''', """hello""" are the same.

In [22]:
print('hello')
print("hello")
print('''hello''')
print("""hello""")

hello
hello
hello
hello


In [25]:
#be careful of number and string

15 #this is an integer number
"15" #this is a string

print(15)
print("15")

15
15


We can display a string literal with the print() function. Alternatively, we can assign a string to a variable, then print the variable

In [90]:
print('hello')

s = 'hello'
print(s)

print('hello')
hello = 100
print(hello)

hello
hello
hello
100


### Escape sequence
- To insert characters that are illegal in a string, we need to use `escape sequence`.

- An escape sequence is a backslash `\` followed by the character you want to insert.

- Some common escape sequences in Python are listed below:

| Escape sequence  |   Description    |
| --------: | :------------- |
| `\'` | Single quote       |
|       `\"` | Double quotes       |
|       `\\` | Backslash(\\)    |
|       `\n` | New Line |
|       `\t` | Tab       |
|       `\b` | Backspace       |
|       `\N{name}` | Character named `name` in the Unicode database       |

Try the examples below to see how to use escape sequence in a program

In [107]:
print("This is a grinning face: \N{grinning face}")
print("I want to print a double quotes \"")
print("I want to print a single quote \'")

#we can use single quote in a string inside a double quote
print("we can print single quote ' directly inside \"")

#we can use double quote in a string inside a single quote
print('we can print double quotes " directly inside \'')

print('a\tb\tc') #\t means Tab key on your keyboard
print('a\nb\nc') #\n will start a new line
print('abc\b') #\b means Backspace, so c will be removed

print('''I
love
programming''') #we can use triple quotes to enclose multi-line string

This is a grinning face: 😀
I want to print a double quotes "
I want to print a single quote '
we can print single quote ' directly inside "
we can print double quotes " directly inside '
a	b	c
a
b
c
ab
I
love
programming


**Summary of escape sequence**

1. To print a simple string, you can use single quote, double quotes, and triple quotes
2. If you want to print `'` inside `'`, you need to use escape sequence `\'`
3. If you want to print `"` inside `"`, you need to use escape sequence `\"`
4. But you can print `'` directly inside `"`. Similarly, you can print `"` inside `'`
5. Triple quotes is usually used to print multiple lines.

**Exercise** Print a cool multi-line string below.

In [108]:
# YOUR CODE HERE
print('''
    7------8
   /|     /|
  3------4 |
  | |    | |
  | 5----|-6
  |/     |/
  1------2
  ''')


    7------8
   /|     /|
  3------4 |
  | |    | |
  | 5----|-6
  |/     |/
  1------2
  


```{seealso}

Compare you answer with the ASCII arts below:
- Star Wars via Telnet: http://asciimation.co.nz/
- bashplotlib: https://github.com/glamp/bashplotlib
```

**Pitfall**

One common error for beginners is they often mix `string` up with `variable name`

In [109]:
# Difference between string and variable name
hello = 5
s = "hello"

print(hello) #this hello is a variable whose value is an integer 5
print("hello") #this "hello" is a string

5
hello


## User Input and Output

**Input**

How to let the user input a value at *runtime*,  i.e., as the program executes?

We can use the method `input()`:
- Syntax: `input([prompt])`
- `prompt` is the string we wish to display on the screen. It is optional.
- There is no need to delimit the input string by quotation marks.
- Simply press `enter` after typing a string.

In [114]:
x = input('Please input a number: ')
x

Please input a number:  10


'10'

In [111]:
x = input()
x

 5


'5'

From the examples above, we can see that
- The `input` method prints its argument, if any, as a [prompt](https://en.wikipedia.org/wiki/Command-line_interface#Command_prompt).  
- It takes user's input and *returns* it as a string, not a number!!

**Output**

How to let your program outputs some data?

We can use the method `print()`:
- Syntax: `print()`
- The `print()` function prints the specified message to the screen, or other standard output device.

The message can be a string, or any other object, the object will be converted into a string before written to the screen.

In [1]:
print("Hello") #print one string
print("Hello,","how are you?") #print two strings
x=10
print("the value of x is:", x) #print one string and one integer

Hello
Hello, how are you?
the value of x is: 10


**Parameters of print()**

print(object(s), sep=separator, end=end)
- `object(s)`	    Any object, and as many as you like. Will be converted to string before printed
- `sep='separator'`	Optional. Specify how to separate the objects, if there is more than one. Default is '  '
- `end='end'`	        Optional. Specify what to print at the end. Default is '\n' (new line)

In [2]:
print("I","love","CS1302!") #By default, strings are separated by space ' '

#change the separator to other characters
print("I","love","CS1302!", sep='-')
print("I","love","CS1302!", sep='*')

I love CS1302!
I-love-CS1302!
I*love*CS1302!


In [3]:
#by default, the ending character is `\n` which starts a new line
print("I")
print("love")
print("CS1302")

#compare the difference
print("I", end='')
print("love", end='')
print("CS1302!")

print("I", end='*')
print("love", end='*')
print("CS1302!")

#what does the following code do?
print() #it prints an empty line

print("I", end='')
print()
print("love", end='')
print()
print("CS1302!")

I
love
CS1302
IloveCS1302!
I*love*CS1302!

I
love
CS1302!


**Exercise** Explain whether the following code prints `'My name is Python'`. Does `print` return a value? 

In [4]:
print('My name is', print('Python'))

Python
My name is None


Answer: print() function returns None.  The None keyword is used to define a null value, or no value at all.
It returns a value! It is None.

## Type Conversion

Python defines type conversion functions to directly convert one data type to another which is useful in day-to-day and competitive programming.

There are two types of Type Conversion in Python:
- Implicit Type Conversion
- Explicit Type Conversion

**Implicit Type Conversion**

In Implicit type conversion, Python automatically converts one data type to another data type. This process doesn't need any user involvement.

Let's see an example where Python promotes the conversion of the lower data type (integer) to the higher data type (float) to avoid data loss.

In [116]:
# Converting integer to float
num_int = 123
num_flo = 1.23

num_new = num_int + num_flo

print("datatype of num_int:",type(num_int))
print("datatype of num_flo:",type(num_flo))

print("Value of num_new:",num_new)
print("datatype of num_new:",type(num_new))

datatype of num_int: <class 'int'>
datatype of num_flo: <class 'float'>
Value of num_new: 124.23
datatype of num_new: <class 'float'>


In the above program,

- We add two variables `num_int` and `num_flo`, storing the value in `num_new`.
- We will look at the data type of all three objects respectively.
- In the output, we can see the data type of `num_int` is an integer while the data type of `num_flo` is a float.
- Also, we can see the `num_new` has a float data type because Python always converts smaller data types to larger data types to avoid the loss of data.

**Note that the data type of a variable, it actually means the data type of the value the variable is holding**

Now, let's try adding a string and an integer, and see how Python deals with it.

In [117]:
# Addition of string(higher) data type and integer(lower) datatype
num_int = 123
num_str = "456"

print("Data type of num_int:",type(num_int))
print("Data type of num_str:",type(num_str))

print(num_int+num_str)

Data type of num_int: <class 'int'>
Data type of num_str: <class 'str'>


TypeError: unsupported operand type(s) for +: 'int' and 'str'

In the above program,

- We add two variables `num_int` and `num_str`.
- As we can see from the output, we got `TypeError`. Python is not able to use Implicit Conversion in such conditions.
- However, Python has a solution for these types of situations which is known as Explicit Conversion.

**Explicit Type Conversion**

In Explicit Type Conversion in Python, the data type is manually changed by the user as per their requirement. Various forms of explicit type conversion are explained below:
 

1. `int()`: This function converts any data type to integer.
2. `float()`: This function converts any data type to a floating-point number
3. `str()`: This function converts any data type to a string

This type of conversion is also called typecasting because the user casts (changes) the data type of the objects.

In [118]:
x = 5 # x is an integer
y = 1.2 #y is a floating point number
z = "100" #z is a string

#example of int()
y_int = int(y)
z_int = int(z)
print(type(y_int))
print(type(z_int))

#example of float()
x_float = float(x)
z_float = float(z)
print(type(x_float))
print(type(z_float))

#example of str()
x_str = str(x)
y_str = str(y)
print(type(x_str))
print(type(y_str))

<class 'int'>
<class 'int'>
<class 'float'>
<class 'float'>
<class 'str'>
<class 'str'>


Note that the input of these functions must be valid, see examples below

In [5]:
x = '123abc'

print(int(x)) #this is wrong
print(float(x)) #this is wrong

ValueError: invalid literal for int() with base 10: '123abc'

In [6]:
x = '123'

print(int(x)) #this is correct
print(float(x)) #this is correct

123
123.0


## String Formatting

How to format a string to the desired format?
- Syntax: `string.format(value1, value2...)`
- The `format()` method formats the specified value(s) and insert them inside the string's placeholder.
- The `format()` method returns the formatted string.
- `value1, value2...`	Required. One or more values that should be formatted and inserted in the string.

**Example 1**

In [4]:
x = 10000/3
print('x ≈',x)
print('x ≈ {:.2f} (rounded to 2 decimal places)'.format(x))
x

x ≈ 3333.3333333333335
x ≈ 3333.33 (rounded to 2 decimal places)


3333.3333333333335

- `{:.2f}` is a place holder
- that gets replaced by a string 
- that represents the argument `x` of `format` 
- according to the [format specification][fspec] `.2f`, i.e.,  
  a decimal floating point number rounded to 2 decimal places.
  
[repf]: https://docs.python.org/3/library/string.html#format-string-syntax
[fspec]: https://docs.python.org/3/library/string.html#format-specification-mini-language

**Exercise** Play with the following widget to learn the effect of different format specifications. In particular, print `10000/3` as `3,333.33`.

In [5]:
from ipywidgets import interact


@interact(
    x="10000/3",
    align={"None": "", "<": "<", ">": ">", "=": "=", "^": "^"},
    sign={"None": "", "+": "+", "-": "-", "SPACE": " "},
    width=(0, 20),
    grouping={"None": "", "_": "_", ",": ","},
    precision=(0, 20),
)
def print_float(x, sign, align, grouping, width=0, precision=2):
    format_spec = (
        f"{{:{align}{sign}{'' if width==0 else width}{grouping}.{precision}f}}"
    )
    print("Format spec:", format_spec)
    print("x ≈", format_spec.format(eval(x)))

interactive(children=(Text(value='10000/3', description='x'), Dropdown(description='sign', options={'None': ''…

**Example 2**

String formatting is useful for different data types other than `float`.  
E.g., consider the following program that prints a time specified by some variables.

In [6]:
# Some specified time
hour = 12
minute = 34
second = 56

print("The time is " + str(hour) + ":" + str(minute) + ":" + str(second)+".")

The time is 12:34:56.


Imagine you have to show also the date in different formats.  
The code can become very hard to read/write because 
- the message is a concatenation of multiple strings and
- the integer variables need to be converted to strings.

Omitting `+` leads to syntax error. Removing `str` as follows also does not give the desired format.

In [7]:
print("The time is ", hour, ":", minute, ":", second, ".")  # note the extra spaces

The time is  12 : 34 : 56 .


To make the code more readable, we can use the `format` function as follows.

In [8]:
message = "The time is {}:{}:{}."
print(message.format(hour, minute, second))

The time is 12:34:56.


```{note}

- We can have multiple *place-holders* `{}` inside a string.
- We can then provide the contents (any type: numbers, strings..) using the `format` function, which
- substitutes the place-holders by the function arguments from left to right.
```

According to the [string formatting syntax](https://docs.python.org/3/library/string.html#format-string-syntax), we can change the order of substitution using 
- indices *(0 is the first item)* or 
- names inside the placeholder `{}`:

In [9]:
print("You should {0} {1} what I say instead of what I {0}.".format("do", "only"))
print("The surname of {first} {last} is {last}.".format(first="John", last="Doe"))

You should do only what I say instead of what I do.
The surname of John Doe is Doe.


**Example 3**

You can even put variables inside the format specification directly and have a nested string formatting.

In [10]:
align, width = "^", 5
print(f"{{:*{align}{width}}}".format(x))  # note the syntax f"..."

3333.3333333333335


In the above, `f"..."` is special syntax for [formatted string](https://docs.python.org/3/library/string.html#format-string-syntax).

**Exercise** Play with the following widget to learn more about the formating specification.  
1. What happens when `align` is none but `fill` is `*`?
1. What happens when the `expression` is a multi-line string?

In [11]:
from ipywidgets import interact


@interact(
    expression=r"'ABC'",
    fill="*",
    align={"None": "", "<": "<", ">": ">", "=": "=", "^": "^"},
    width=(0, 20),
)
def print_objectt(expression, fill, align="^", width=10):
    format_spec = f"{{:{fill}{align}{'' if width==0 else width}}}"
    print("Format spec:", format_spec)
    print("Print:", format_spec.format(eval(expression)))

interactive(children=(Text(value="'ABC'", description='expression'), Text(value='*', description='fill'), Drop…

## Error

In addition to writing code, a programmer spends significant time in *debugging* code that contains errors.

**Can an error be automatically detected by the computer?**

In [30]:
#example of logical error
x = 5
y = 0
print(x/y)

ZeroDivisionError: division by zero

- You have just seen an example of *logical error*, which is due to an error in the logic.  
- The ability to debug or even detect such error is, unfortunately, beyond python's intelligence.

Other kinds of error may be detected automatically.  
As an example, note that we can omit `+` for string concatenation, but we cannot omit it for integer summation:

In [12]:
print('Skipping + for string concatenation')
'4''5' '6'

Skipping + for string concatenation


'456'

In [13]:
print('Skipping + for integer summation')
4 5 6

SyntaxError: invalid syntax (3402839676.py, line 2)

Python interpreter detects the bug and raises a *syntax* error.

**Why Syntax error can be detected automatically?  
Why is the print statement before the error not executed?**

- The Python interpreter can detect syntax error even before executing the code because
- the interpreter simply fails to translate the code to lower-level executable code.

The following code raises a different kind of error.

In [31]:
print("Evaluating '4' + '5' + 6")
'4' + '5' + 6  # summing string with integer

Evaluating '4' + '5' + 6


TypeError: can only concatenate str (not "int") to str

**Why Python throws a TypeError when evaluating `'4' + '5' + 6`?**

There is no implementation of `+` operation on a value of type `str` and a value of type `int`. 

- Unlike the syntax error, the Python interpreter can only detect a type error at runtime (when executing the code.) 
- Hence, such an error is called a *runtime error*.


**Do other languages work in the same way?**

The following cell runs the corresponding javascript code using a cell magic:

In [14]:
%%javascript
let x = '4' + '5' + 6;
element.append(x + ' ' + typeof(x));
// no error because 6 is converted to a str implicitly

<IPython.core.display.Javascript object>

In [15]:
%%javascript
let x = '4' * '5' * 6;
element.append(x + ' ' + typeof(x));
// no error because 4 and 5 are converted to numbers implicitly

<IPython.core.display.Javascript object>

Javascript is called is a *weakly-typed* language because it forces a type conversion to avoid a type error.

Python is a [*strongly-and-dynamically-typed*](https://en.wikipedia.org/wiki/Strong_and_weak_typing) language:
- *Strongly-typed*: Python does not force a type conversion to avoid a type error.
- *Dynamically-typed*: Python checks data type only at runtime after translating the code to machine code.

In comparison, C/C++ and Java are *statically-typed* languages that checks data type during compilation.

In [20]:
try:
    !gcc 456.cpp
except Error:
    print('Cannot run shell command.')

456.cpp: In function ‘int main()’:
456.cpp:4:22: error: invalid operands of types ‘const char [2]’ and ‘const char [2]’ to binary ‘operator+’
    4 |     std::cout << "4" + "5" + 6 << std::endl;
      |                  ~~~ ^ ~~~
      |                  |     |
      |                  |     const char [2]
      |                  const char [2]


In [21]:
try:
    !javac 456.java
except Error:
    print('Cannot run shell command.')

456.java:3: error: bad operand types for binary operator '*'
        System.out.println("4" * "5" * 6);
                               ^
  first type:  String
  second type: String
1 error


```{note}

A weakly-typed language may seem more robust, but it can lead to [more logical errors](https://www.oreilly.com/library/view/fluent-conference-javascript/9781449339203/oreillyvideos1220106.html).

- Javascript is [tricky](https://github.com/denysdovhan/wtfjs).
- To improve readability and avoid logical errors, [typescript](https://www.typescriptlang.org/) is a strongly-typed replacement of javascript.
```

**Exercise** Not all the strings can be converted into integers. Try breaking the following code by providing invalid inputs and record them in the subsequent cell. Explain whether the errors are runtime errors.

In [12]:
num1 = input('Please input an integer: ')
num2 = input('Please input another integer: ')
print(num1, '+', num2, 'is equal to', int(num1) + int(num2))

Please input an integer:  abc
Please input another integer:  5


ValueError: invalid literal for int() with base 10: 'abc'

ANSWER: the errors are runtime error